### ISSUE TITLE
Seeing EncodingError with HPOneViewException (unable to pickle class

### ISSUE THREAD

**Issue by eamonnotoole** [HewlettPackard/python-hpOneView#320]
### Scenario/Intent
I'm running python-hpOneView as a Celery task.  I have 3 containers, one running Celery/python-hpOneView, another running an application that fires-off the Celery task, and a third running Redis.  I want to process HPOneViewExceptions to add the exception string to a database.  This is what I see in Celery logs:

celery_1     | [2017-10-02 09:49:13,683: ERROR/ForkPoolWorker-9] Task ncsappman.servers.remove_sp_from_server[d9cf7888-6fe3-4c17-af2f-81bcb647eec1] raised unexpected: HPOneViewException(u'Resource not found, ID/URI = ed84ee6f-5c0f-43e8-99f7-8520189cb149.', {u'errorSource': None, u'recommendedActions': [u'Confirm resource ID/URI and try the call again.'], u'nestedErrors': [], u'errorCode': u'RESOURCE_NOT_FOUND', u'details': u'The requested resource not found, ID = ed84ee6f-5c0f-43e8-99f7-8520189cb149.', u'message': u'Resource not found, ID/URI = ed84ee6f-5c0f-43e8-99f7-8520189cb149.', u'data': {}})
celery_1     | Traceback (most recent call last):
celery_1     |   File "/usr/local/lib/python2.7/dist-packages/celery/app/trace.py", line 374, in trace_task
celery_1     |     R = retval = fun(*args, **kwargs)
celery_1     |   File "/usr/local/lib/python2.7/dist-packages/celery/app/trace.py", line 629, in __protected_call__
celery_1     |     return self.run(*args, **kwargs)
celery_1     |   File "/app/ncsappman/servers.py", line 153, in remove_sp_from_server
celery_1     |     client.server_profiles.delete(content['serverProfileUri'])
celery_1     |   File "/python-hpOneView/hpOneView/resources/servers/server_profiles.py", line 121, in delete
celery_1     |     return self._client.delete(resource=resource, timeout=timeout)
celery_1     |   File "/python-hpOneView/hpOneView/resources/resource.py", line 259, in delete
celery_1     |     task, body = self._connection.delete(uri, custom_headers=custom_headers)
celery_1     |   File "/python-hpOneView/hpOneView/connection.py", line 359, in delete
celery_1     |     return self.__do_rest_call('DELETE', uri, {}, custom_headers=custom_headers)
celery_1     |   File "/python-hpOneView/hpOneView/connection.py", line 441, in __do_rest_call
celery_1     |     raise HPOneViewException(body)
celery_1     | HPOneViewException: (u'Resource not found, ID/URI = ed84ee6f-5c0f-43e8-99f7-8520189cb149.', {u'errorSource': None, u'recommendedActions': [u'Confirm resource ID/URI and try the call again.'], u'nestedErrors': [], u'errorCode': u'RESOURCE_NOT_FOUND', u'details': u'The requested resource not found, ID = ed84ee6f-5c0f-43e8-99f7-8520189cb149.', u'message': u'Resource not found, ID/URI = ed84ee6f-5c0f-43e8-99f7-8520189cb149.', u'data': {}})
celery_1     | [2017-10-02 09:49:13,692: ERROR/MainProcess] Task handler raised error: <MaybeEncodingError: Error sending result: '"(1, <ExceptionInfo: HPOneViewException(u'Resource not found, ID/URI = ed84ee6f-5c0f-43e8-99f7-8520189cb149.', {u'errorSource': None, u'recommendedActions': [u'Confirm resource ID/URI and try the call again.'], u'nestedErrors': [], u'errorCode': u'RESOURCE_NOT_FOUND', u'details': u'The requested resource not found, ID = ed84ee6f-5c0f-43e8-99f7-8520189cb149.', u'message': u'Resource not found, ID/URI = ed84ee6f-5c0f-43e8-99f7-8520189cb149.', u'data': {}})>, None)"'. Reason: ''PicklingError("Can\'t pickle <class \'hpOneView.exceptions.HPOneViewException\'>: it\'s not the same object as hpOneView.exceptions.HPOneViewException",)''.>
celery_1     | Traceback (most recent call last):
celery_1     |   File "/usr/local/lib/python2.7/dist-packages/billiard/pool.py", line 362, in workloop
celery_1     |     put((READY, (job, i, result, inqW_fd)))
celery_1     |   File "/usr/local/lib/python2.7/dist-packages/billiard/queues.py", line 366, in put
celery_1     |     self.send_payload(ForkingPickler.dumps(obj))
celery_1     |   File "/usr/local/lib/python2.7/dist-packages/billiard/reduction.py", line 82, in dumps
celery_1     |     cls(buf, protocol).dump(obj)
celery_1     |   File "/usr/lib/python2.7/pickle.py", line 224, in dump
celery_1     |     self.save(obj)
celery_1     |   File "/usr/lib/python2.7/pickle.py", line 286, in save
celery_1     |     f(self, obj) # Call unbound method with explicit self
celery_1     |   File "/usr/lib/python2.7/pickle.py", line 562, in save_tuple
celery_1     |     save(element)
celery_1     |   File "/usr/lib/python2.7/pickle.py", line 286, in save
celery_1     |     f(self, obj) # Call unbound method with explicit self
celery_1     |   File "/usr/lib/python2.7/pickle.py", line 562, in save_tuple
celery_1     |     save(element)
celery_1     |   File "/usr/lib/python2.7/pickle.py", line 286, in save
celery_1     |     f(self, obj) # Call unbound method with explicit self
celery_1     |   File "/usr/lib/python2.7/pickle.py", line 562, in save_tuple
celery_1     |     save(element)
celery_1     |   File "/usr/lib/python2.7/pickle.py", line 286, in save
celery_1     |     f(self, obj) # Call unbound method with explicit self
celery_1     |   File "/usr/lib/python2.7/pickle.py", line 562, in save_tuple
celery_1     |     save(element)
celery_1     |   File "/usr/lib/python2.7/pickle.py", line 331, in save
celery_1     |     self.save_reduce(obj=obj, *rv)
celery_1     |   File "/usr/lib/python2.7/pickle.py", line 419, in save_reduce
celery_1     |     save(state)
celery_1     |   File "/usr/lib/python2.7/pickle.py", line 286, in save
celery_1     |     f(self, obj) # Call unbound method with explicit self
celery_1     |   File "/usr/lib/python2.7/pickle.py", line 649, in save_dict
celery_1     |     self._batch_setitems(obj.iteritems())
celery_1     |   File "/usr/lib/python2.7/pickle.py", line 663, in _batch_setitems
celery_1     |     save(v)
celery_1     |   File "/usr/lib/python2.7/pickle.py", line 331, in save
celery_1     |     self.save_reduce(obj=obj, *rv)
celery_1     |   File "/usr/lib/python2.7/pickle.py", line 400, in save_reduce
celery_1     |     save(func)
celery_1     |   File "/usr/lib/python2.7/pickle.py", line 286, in save
celery_1     |     f(self, obj) # Call unbound method with explicit self
celery_1     |   File "/usr/lib/python2.7/pickle.py", line 753, in save_global
celery_1     |     (obj, module, name))
celery_1     | MaybeEncodingError: Error sending result: '"(1, <ExceptionInfo: HPOneViewException(u'Resource not found, ID/URI = ed84ee6f-5c0f-43e8-99f7-8520189cb149.', {u'errorSource': None, u'recommendedActions': [u'Confirm resource ID/URI and try the call again.'], u'nestedErrors': [], u'errorCode': u'RESOURCE_NOT_FOUND', u'details': u'The requested resource not found, ID = ed84ee6f-5c0f-43e8-99f7-8520189cb149.', u'message': u'Resource not found, ID/URI = ed84ee6f-5c0f-43e8-99f7-8520189cb149.', u'data': {}})>, None)"'. Reason: ''PicklingError("Can\'t pickle <class \'hpOneView.exceptions.HPOneViewException\'>: it\'s not the same object as hpOneView.exceptions.HPOneViewException",)''.

### Environment Details
OV API version 500, python-hpOneView v4

### Steps to Reproduce
[What are the complete steps needed to do in order to reproduce your problem?]

### Expected Result
This is what I expect to see instead of the above:

celery_1     | [2017-10-02 10:21:22,469: ERROR/ForkPoolWorker-9] Task ncsappman.servers.remove_sp_from_server[f2c7f252-50a1-44b8-b190-532a529907e6] raised unexpected: HPOneViewException(u"Resource not found, ID/URI = ed84ee6f-5c0f-43e8-99f7-8520189cb149., {u'errorSource': None, u'recommendedActions': [u'Confirm resource ID/URI and try the call again.'], u'nestedErrors': [], u'errorCode': u'RESOURCE_NOT_FOUND', u'details': u'The requested resource not found, ID = ed84ee6f-5c0f-43e8-99f7-8520189cb149.', u'message': u'Resource not found, ID/URI = ed84ee6f-5c0f-43e8-99f7-8520189cb149.', u'data': {}}",)
celery_1     | Traceback (most recent call last):
celery_1     |   File "/usr/local/lib/python2.7/dist-packages/celery/app/trace.py", line 374, in trace_task
celery_1     |     R = retval = fun(*args, **kwargs)
celery_1     |   File "/usr/local/lib/python2.7/dist-packages/celery/app/trace.py", line 629, in __protected_call__
celery_1     |     return self.run(*args, **kwargs)
celery_1     |   File "/app/ncsappman/servers.py", line 153, in remove_sp_from_server
celery_1     |     client.server_profiles.delete(content['serverProfileUri'])
celery_1     |   File "/python-hpOneView/hpOneView/resources/servers/server_profiles.py", line 121, in delete
celery_1     |     return self._client.delete(resource=resource, timeout=timeout)
celery_1     |   File "/python-hpOneView/hpOneView/resources/resource.py", line 259, in delete
celery_1     |     task, body = self._connection.delete(uri, custom_headers=custom_headers)
celery_1     |   File "/python-hpOneView/hpOneView/connection.py", line 359, in delete
celery_1     |     return self.__do_rest_call('DELETE', uri, {}, custom_headers=custom_headers)
celery_1     |   File "/python-hpOneView/hpOneView/connection.py", line 441, in __do_rest_call
celery_1     |     raise HPOneViewException(body)
celery_1     | HPOneViewException: Resource not found, ID/URI = ed84ee6f-5c0f-43e8-99f7-8520189cb149., {u'errorSource': None, u'recommendedActions': [u'Confirm resource ID/URI and try the call again.'], u'nestedErrors': [], u'errorCode': u'RESOURCE_NOT_FOUND', u'details': u'The requested resource not found, ID = ed84ee6f-5c0f-43e8-99f7-8520189cb149.', u'message': u'Resource not found, ID/URI = ed84ee6f-5c0f-43e8-99f7-8520189cb149.', u'data': {}}

### What is happening

The problem seems to be in the __init__ method of HPOneViewException:

class HPOneViewException(Exception):
    """
    OneView base Exception.

    Attributes:
       msg (str): Exception message.
       oneview_response (dict): OneView rest response.
   """

    def __init__(self, data):
        self.msg = None
        self.oneview_response = None

        if isinstance(data, basestring):
            self.msg = data
        else:
            self.oneview_response = data

            if data and isinstance(data, dict):
                self.msg = data.get('message')

        if self.oneview_response:
            Exception.__init__(self, self.msg, self.oneview_response)
        else:
            Exception.__init__(self, self.msg)

When self.oneview_response exists.  Exception is called with two arguments, when HPOneViewException is defined to take one argument.  The easiest way to fix this, I think, is to do something like the following:

[eamonn@z620-eamonn hpOneView (fix-pickle)]$ git diff exceptions.py
diff --git a/hpOneView/exceptions.py b/hpOneView/exceptions.py
index 5aa5c48..d2f7cd1 100644
--- a/hpOneView/exceptions.py
+++ b/hpOneView/exceptions.py
@@ -78,11 +78,9 @@ class HPOneViewException(Exception):
 
             if data and isinstance(data, dict):
                 self.msg = data.get('message')
+                self.msg += ', %s' % self.oneview_response
 
-        if self.oneview_response:
-            Exception.__init__(self, self.msg, self.oneview_response)
-        else:
-            Exception.__init__(self, self.msg)
+        Exception.__init__(self, self.msg)
 
I've verified that this change will produce the desired output above.  It may be possible to fix by overriding the __reduce__ method, I didn't try that.

**Comment by fgbulsoni** [comment]
> Exception is called with two arguments, when HPOneViewException is defined to take one argument.

Is the issue calling `Exception` with two arguments, or expecting only one in the `__init__ `method?

If the issue is expecting only one argument in __init__, wouldn't adding another param to `HpOneViewException`'s `__init__` be a good way fix it?
Something like:

```Python
(...)
class HPOneViewException(Exception):
    def __init__(self, data, error=None):
        self.msg = None
        self.oneview_response = None
(...)
```
Otherwise, if the issue is specifically with calling `Exception.__init__` using two arguments, then sure, we can likely just get the message and use it as necessary, as suggested.

**Comment by eamonnotoole** [comment]
The issue seems to be number of arguments to the `__init__` method of HPOneViewException.

I had put-up a fix that used an override to `__reduce__`, which fixes the problem while allowing all unit tests to pass.  However adding another argument to the `__init__` method works as well and is simpler.

I've added a couple of unit tests to test pickling.

**Comment by fgbulsoni** [comment]
Closed via #322
